File: apps/syn-cli/src/syn_cli/commands/control.py
"""Execution control commands — pause, resume, cancel, status, stop."""

from __future__ import annotations

from typing import Annotated

import typer

from syn_cli._output import console
from syn_cli.commands._api_helpers import api_get, api_post

app = typer.Typer(
    name="control",
    help="Control running executions",
    no_args_is_help=True,
)

_STATE_COLORS: dict[str, str] = {
    "pending": "dim",
    "running": "blue",
    "paused": "yellow",
    "cancelled": "orange3",
    "interrupted": "orange3",
    "completed": "green",
    "failed": "red",
}


@app.command()
def pause(
    execution_id: Annotated[str, typer.Argument(help="Execution ID to pause")],
    reason: Annotated[str | None, typer.Option("--reason", "-r", help="Reason for pausing")] = None,
) -> None:
    """Pause a running execution at the next yield point."""
    data = api_post(
        f"/api/executions/{execution_id}/pause",
        json={"reason": reason} if reason else None,
    )

    console.print(f"[green]Pause signal sent for execution {execution_id}[/green]")
    console.print(f"  State: {data.get('state', 'unknown')}")
    if data.get("message"):
        console.print(f"  Message: {data['message']}")


@app.command()
def resume(
    execution_id: Annotated[str, typer.Argument(help="Execution ID to resume")],
) -> None:
    """Resume a paused execution."""
    data = api_post(f"/api/executions/{execution_id}/resume")

    console.print(f"[green]Resume signal sent for execution {execution_id}[/green]")
    console.print(f"  State: {data.get('state', 'unknown')}")


@app.command()
def cancel(
    execution_id: Annotated[str, typer.Argument(help="Execution ID to cancel")],
    reason: Annotated[
        str | None, typer.Option("--reason", "-r", help="Reason for cancelling")
    ] = None,
    force: Annotated[bool, typer.Option("--force", "-f", help="Skip confirmation prompt")] = False,
) -> None:
    """Cancel a running or paused execution."""
    if not force:
        confirm = typer.confirm(f"Are you sure you want to cancel execution {execution_id}?")
        if not confirm:
            console.print("[dim]Cancelled.[/dim]")
            raise typer.Exit(0)

    data = api_post(
        f"/api/executions/{execution_id}/cancel",
        json={"reason": reason} if reason else None,
    )

    console.print(f"[green]Cancel signal sent for execution {execution_id}[/green]")
    console.print(f"  State: {data.get('state', 'unknown')}")


def _render_interrupted_detail(execution_id: str) -> None:
    """Fetch and display additional context for interrupted executions."""
    try:
        detail = api_get(f"/api/executions/{execution_id}")
        if detail.get("error_message"):
            console.print(f"  Reason: {detail['error_message']}")
        if detail.get("completed_at"):
            console.print(f"  Interrupted at: {detail['completed_at']}")
        if detail.get("git_sha"):
            console.print(f"  Git SHA: {detail['git_sha']}")
    except SystemExit:
        pass


@app.command()
def status(
    execution_id: Annotated[str, typer.Argument(help="Execution ID to check")],
) -> None:
    """Get current execution control state."""
    data = api_get(f"/api/executions/{execution_id}/state")
    state = data.get("state", "unknown")
    color = _STATE_COLORS.get(state, "white")

    console.print(f"Execution: {execution_id}")
    console.print(f"State: [{color}]{state}[/{color}]")

    if state == "interrupted":
        _render_interrupted_detail(execution_id)


@app.command()
def inject(
    execution_id: Annotated[str, typer.Argument(help="Execution ID")],
    message: Annotated[str, typer.Option("--message", "-m", help="Message to inject")],
) -> None:
    """Inject a message into a running execution."""
    api_post(f"/executions/{execution_id}/inject", json={"message": message})
    console.print(f"[green]Message injected into execution {execution_id}[/green]")


@app.command()
def stop(
    execution_id: Annotated[str, typer.Argument(help="Execution ID to stop")],
    reason: Annotated[
        str | None, typer.Option("--reason", "-r", help="Reason for stopping")
    ] = None,
    force: Annotated[bool, typer.Option("--force", "-f", help="Skip confirmation prompt")] = False,
) -> None:
    """Forcefully stop a running execution via SIGINT.

    Sends a cancel signal that causes the engine to interrupt the Claude CLI
    process via SIGINT and capture partial output as an interrupted execution.
    """
    stop_reason = reason or "Stopped by user via syn stop"

    if not force:
        confirm = typer.confirm(f"Are you sure you want to stop execution {execution_id}?")
        if not confirm:
            console.print("[dim]Aborted.[/dim]")
            raise typer.Exit(0)

    data = api_post(
        f"/api/executions/{execution_id}/cancel",
        json={"reason": stop_reason},
    )

    console.print(f"[orange3]Stop signal sent for execution {execution_id}[/orange3]")
    console.print(f"  State: {data.get('state', 'unknown')}")
    if data.get("message"):
        console.print(f"  Message: {data['message']}")

File: apps/syn-cli-node/src/commands/control.ts
/**
 * Execution control commands — pause, resume, cancel, status, inject, stop.
 * Port of apps/syn-cli/src/syn_cli/commands/control.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { api, unwrap } from "../client/typed.js";
import type { components } from "../generated/api-types.js";
import { print, printError, printDim } from "../output/console.js";
import { style, GREEN, YELLOW } from "../output/ansi.js";
import { formatStatus } from "../output/format.js";

type ControlResponse = components["schemas"]["ControlResponse"];
type StateResponse = components["schemas"]["StateResponse"];

/**
 * Render the body of a control response.
 *
 * pause, resume, cancel and stop all queue a signal and return immediately, so
 * `state` is the execution's state from *before* the signal was queued. Printing
 * it as a bare `State:` under "signal sent" reads as "the command did nothing" —
 * which is how a live cancel gets reported as a stuck execution. Label it for
 * what it is, and always surface the server's own message, which says what was
 * actually queued.
 */
function printSignalQueued(data: ControlResponse): void {
  print(`  State before signal: ${data.state}`);
  if (data.message) print(`  Message: ${data.message}`);
  printDim("  Asynchronous: run `syn control status <execution-id>` for the new state.");
}

function reqId(parsed: ParsedArgs): string {
  const id = parsed.positionals[0];
  if (!id) {
    printError("execution-id is required");
    printDim("Hint: run `syn execution list` to find an execution ID.");
    throw new CLIError("Missing argument", 1);
  }
  return id;
}

const pauseCommand: CommandDef = {
  name: "pause",
  description: "Pause a running execution at the next yield point",
  args: [{ name: "execution-id", description: "Execution ID to pause", required: true }],
  options: { reason: { type: "string", short: "r", description: "Reason for pausing" } },
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    const reason = parsed.values["reason"] as string | undefined;
    const data = unwrap<ControlResponse>(
      await api.POST("/executions/{execution_id}/pause", {
        params: { path: { execution_id: id } },
        ...(reason ? { body: { reason } } : {}),
      }),
      "Pause execution",
    );
    print(style(`Pause signal sent for execution ${id}`, GREEN));
    printSignalQueued(data);
  },
};

const resumeCommand: CommandDef = {
  name: "resume",
  description: "Resume a paused execution",
  args: [{ name: "execution-id", description: "Execution ID to resume", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    const data = unwrap<ControlResponse>(
      await api.POST("/executions/{execution_id}/resume", {
        params: { path: { execution_id: id } },
      }),
      "Resume execution",
    );
    print(style(`Resume signal sent for execution ${id}`, GREEN));
    printSignalQueued(data);
  },
};

const cancelCommand: CommandDef = {
  name: "cancel",
  description: "Cancel a running or paused execution",
  args: [{ name: "execution-id", description: "Execution ID to cancel", required: true }],
  options: {
    reason: { type: "string", short: "r", description: "Reason for cancelling" },
    force: { type: "boolean", short: "f", description: "Skip confirmation prompt", default: false },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    if (parsed.values["force"] !== true) {
      printError(`Use --force to confirm cancelling execution ${id}`);
      throw new CLIError("Confirmation required", 1);
    }
    const reason = parsed.values["reason"] as string | undefined;
    const data = unwrap<ControlResponse>(
      await api.POST("/executions/{execution_id}/cancel", {
        params: { path: { execution_id: id } },
        ...(reason ? { body: { reason } } : {}),
      }),
      "Cancel execution",
    );
    print(style(`Cancel signal sent for execution ${id}`, GREEN));
    printSignalQueued(data);
  },
};

const statusCommand: CommandDef = {
  name: "status",
  description: "Get current execution control state",
  args: [{ name: "execution-id", description: "Execution ID to check", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    const data = unwrap<StateResponse>(
      await api.GET("/executions/{execution_id}/state", {
        params: { path: { execution_id: id } },
      }),
      "Get execution state",
    );
    print(`Execution: ${id}`);
    print(`State: ${formatStatus(data.state)}`);
  },
};

const injectCommand: CommandDef = {
  name: "inject",
  description: "Inject a message into a running execution",
  args: [{ name: "execution-id", description: "Execution ID", required: true }],
  options: {
    message: { type: "string", short: "m", description: "Message to inject" },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    const message = parsed.values["message"] as string | undefined;
    if (!message) { printError("Missing --message"); throw new CLIError("Missing option", 1); }
    unwrap(
      await api.POST("/executions/{execution_id}/inject", {
        params: { path: { execution_id: id } },
        body: { message, role: "user" },
      }),
      "Inject message",
    );
    print(style(`Message injected into execution ${id}`, GREEN));
  },
};

const stopCommand: CommandDef = {
  name: "stop",
  description: "Forcefully stop a running execution via SIGINT",
  args: [{ name: "execution-id", description: "Execution ID to stop", required: true }],
  options: {
    reason: { type: "string", short: "r", description: "Reason for stopping" },
    force: { type: "boolean", short: "f", description: "Skip confirmation prompt", default: false },
  },
  handler: async (parsed: ParsedArgs) => {
    const id = reqId(parsed);
    if (parsed.values["force"] !== true) {
      printError(`Use --force to confirm stopping execution ${id}`);
      throw new CLIError("Confirmation required", 1);
    }
    const reason = (parsed.values["reason"] as string | undefined) ?? "Stopped by user via syn stop";
    const data = unwrap<ControlResponse>(
      await api.POST("/executions/{execution_id}/cancel", {
        params: { path: { execution_id: id } },
        body: { reason },
      }),
      "Stop execution",
    );
    print(style(`Stop signal sent for execution ${id}`, YELLOW));
    printSignalQueued(data);
  },
};

export const controlGroup = new CommandGroup(
  "control",
  "Control running executions",
  "`pause`, `resume`, `cancel` and `stop` are asynchronous: they queue a signal " +
    "for the executor and return straight away. The state they print is the state " +
    "read *before* the signal was queued, and `syn control status` may report that " +
    "same state for a short while afterwards. That is not a refused command - " +
    "poll `syn control status` until it changes.",
);
controlGroup
  .command(pauseCommand)
  .command(resumeCommand)
  .command(cancelCommand)
  .command(statusCommand)
  .command(injectCommand)
  .command(stopCommand);
